rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,556 @@
+                            sub_digraph,
+                            tokens,
+                            todo_nodes,
+                        );
+                        steps += 1;
+                        found = true;
+                        break;
+                    }
+                }
+                // If none found, look for cycle in digraph which will result in
+                // an unhappy node. Look for a predecessor and add node and pred
+                // to swap_edges and do swap
+                if !found {
+                    let cycle: Vec<Edge> = find_cycle(&*digraph, Some(todo_node));
+                    let unhappy_node = cycle[0].0;
+                    let mut found = false;
+                    let di2 = &mut digraph.clone();
+                    for predecessor in di2.neighbors_directed(unhappy_node, Incoming) {
+                        if predecessor != unhappy_node {
+                            swap_edges.push((unhappy_node, predecessor));
+                            self.swap(
+                                unhappy_node,
+                                predecessor,
+                                digraph,
+                                sub_digraph,
+                                tokens,
+                                todo_nodes,
+                            );
+                            steps += 1;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if !found {
+                        panic!("unexpected stop")
+                    }
+                }
+            }
+        }
+        if !todo_nodes.is_empty() {
+            panic!("got todo nodes");
+        }
+        swap_edges
+    }
+
+    fn swap(
+        &mut self,
+        node1: NodeIndex,
+        node2: NodeIndex,
+        digraph: &mut StableGraph<(), (), Directed>,
+        sub_digraph: &mut StableGraph<(), (), Directed>,
+        tokens: &mut HashMap<NodeIndex, NodeIndex>,
+        todo_nodes: &mut Vec<NodeIndex>,
+    ) {
+        // Get token values for the 2 nodes and remove them
+        let token1 = tokens.remove(&node1);
+        let token2 = tokens.remove(&node2);
+
+        // Swap the token edge values
+        if let Some(t2) = token2 {
+            tokens.insert(node1, t2);
+        }
+        if let Some(t1) = token1 {
+            tokens.insert(node2, t1);
+        }
+        // For each node, remove the (node, successor) from digraph and
+        // sub_digraph. Then add new token edges back in.
+        for node in [node1, node2] {
+            let mut edge_nodes = vec![];
+            for successor in digraph.neighbors_directed(node, Outgoing) {
+                edge_nodes.push((node, successor));
+            }
+            for (edge_node1, edge_node2) in edge_nodes {
+                let edge = digraph.find_edge(edge_node1, edge_node2).unwrap();
+                digraph.remove_edge(edge);
+            }
```

Comment:
Is there a reason to use `edge_nodes` as a vec here? Can't we just do this as:
```rust
for edge_node2 in digraph.neighbors_directed(node, Outgoing) {
    let edge = digraph.find_edge(node, edge_node2).unwrap();
    digraph.remove_edge(edge);
}
```

This will avoid any memory allocations which will be much faster.

Reply:
Done in https://github.com/Qiskit/rustworkx/pull/765/commits/2d74c45a8e8c6e763cd7ce32034d9f094f991661